feat(dart/transform): Run the transformer on benchmarks

kegluneq opened the pull request:
- Remove explicit generation of reflection information in benchmark code
  and generate it with the transformer.
- Update transformer to handle cases in benchmarks
- Gracefully handle log calls before init
- Allow multiple transformer entry points
- Update ReflectionRemover to use AssetReader
- Do not import reflection_capabilities.ng_deps.dart

yjbanov commented on modules/angular2/src/transform/bind_generator/generator.dart:
```diff
@@ -22,12 +22,24 @@ Future<String> createNgSetters(AssetReader reader, AssetId entryPoint) async {
+    switch (prop) {
+      case 'if':
+      case 'of':
+        setters.add(''' '$sProp': (o, v) => o['$sProp'] = v ''');
```
Do you know why we need these special cases? The actual setters in `If` and `For` directives are `condition` and `iterableChanges`. I'm surprised we are reaching this code with `if` and `of`.

kegluneq replied:
I'm not sure - those values are coming out of the compiler, so maybe Tobias would have some insight?

yjbanov replied:
What seems to be happening:

```
<div *if="expr">Hello</div>
```

Is converted to

```
<template if [if]="expr">
  <div>Hello</div>
</template>
```

And the compiler thinks that `[if]` might be binding to an element property `if`. This is a bug. Does it work if you return `null`? If so, let's return null, log an issue, and leave a TODO here to remove later.

kegluneq replied:
Since users could try to bind to `if` (or other keywords) in their templates, maybe we should `throw` and leave this in?

yjbanov replied:
Do you mean:

```
case 'if':
case 'of':
  throw 'invalid setter';
```

Or 

```
case 'if':
case 'of':
  setters.add(''' '$sProp': (o, v) { throw 'invalid setter'; }''');
```

?

kegluneq replied:
the second, executing the setter should `throw`, explaining what went wrong

yjbanov replied:
Let's do it as a temporary workaround, but let's also think about how we can avoid generating this code completely. It seems there are two possible situations, the one above, and this:

```
<my-component [if]="whatever"></>
```

In the case when we are dealing with a `@Viewport` directive the compiler should know better and not ask for a setter in the first place. In the second case it's a programmer error and we should throw early.

So it seems to me that the ultimate solution is to fix the compiler and then throw eagerly rather than from inside the generated setter code.

yjbanov commented on modules/angular2/src/transform/template_compiler/generator.dart:
```diff
@@ -61,17 +61,37 @@ Future<String> processTemplates(AssetReader reader, AssetId entryPoint) async {
+    switch (prop) {
+      case 'if':
+      case 'of':
+        return ''' '$sProp': (o) => o['$sProp'] ''';
```
If I understand correctly, this is for expressions, so we shouldn't hit `if` and `of` cases at all. I think it's safe to throw immediately with a descriptive message, such as "$prop is a Dart reserved word and cannot be used as getter/setter name".

yjbanov commented on modules/angular2/src/transform/template_compiler/generator.dart:
```diff
@@ -61,17 +61,37 @@ Future<String> processTemplates(AssetReader reader, AssetId entryPoint) async {
+    switch (prop) {
+      case 'if':
+      case 'of':
+        return ''' '$sProp': (o, v) => o['$sProp'] = v ''';
```
ditto

yjbanov commented:
LGTM if LGT Travis

Minor AIs:
- throw lazily from illegal setters `if` and `of`
- log an issue to fix the compiler (it should not ask for setters on `@Viewport` bindings)
- add TODO (with reference to the issue) to throw eagerly

kegluneq commented:
It looks like Travis is happy, so that's good! Now that things are all passing, I've removed the necessity for the `ngstatic` parameter (& will remove the associated gulp targets before submitting)

Responses to your bullets & a question below:

1. throw lazily from illegal setters if and of
Done, and we now check against `analyzer`'s list of Dart keywords.

2. log an issue to fix the compiler (it should not ask for setters on @Viewport bindings)
I'm not 100% certain this is a `Compiler` issue, but maybe you can clarify for me (detail below)

3. add TODO (with reference to the issue) to throw eagerly
Will do once we determine `#2` above

#### Compiler "bug" detail
We're not currently running the full compiler, just pieces of the `CompilePipeline`. We set

```
reflector.reflectionCapabilities = new RecordingReflectionCapabilities();
```
[link](https://github.com/angular/angular/blob/master/modules/angular2/src/transform/template_compiler/recording_reflection_capabilities.dart)

And then we know what pieces of the template the compile pipeline tried to access reflectively and generate getters, setters, and methods from there.

I imagine this is still incorrect behavior on the `CompilerPipeline`, but would you expect, for example, in  [app.js][] we should be calling `reflector.setter('if')`?

[app.js]: https://github.com/angular/angular/blob/master/modules/benchmarks/src/naive_infinite_scroll/app.js#L94

yjbanov commented:
In that example we should expect the compiler to be calling `reflector.setter('condition')` because it should know that it is a `@Viewport` directive defined like this:

```dart
@Viewport({
  selector: '[if]',
  bind: {
    'condition': 'if'
  }
})
class If {
  set condition(v) { ... }
}
```

My guess is that right now it is calling _both_ `reflector.setter('condidion')` _and_ `reflector.setter('if')` while only the former is needed.

kegluneq commented:
I don't think that's the case I'm seeing currently -- the generated `if` setter is being generated directly from the `@Template` annotation on App, seen [here](https://github.com/angular/angular/blob/master/modules/benchmarks/src/naive_infinite_scroll/app.js#L94)

Commit:
feat(dart/transform): Mark Compiler as Injectable

Necessary to allow runtime access via the `Injector`

Commit:
feat(dart/transform): Add stub implementations to Html5LibAdapter

Stub out some methods used in the `CompilerPipeline`.

Commit:
fix(dart/transform): Gracefully handle log calls before init

- Lazily create and use a logger that prints instead of `throw`ing.
- Use this logger in unit tests.

Commit:
feat(dart/transform): Allow multiple transformer entry points

- Allow the user to specify multiple entry points to an app.
- Allow the Angular 2 transformer to run without explicit entry points to
generate necessary setters & getters on built-in directives like `For`
and `If`.

Closes #1246

Commit:
feat(dart/transform): Fix handling of Dart keywords

Use `package:analyzer`'s list of Dart keywords to ensure we are properly
reporting usages of Dart keywords as runtime errors.

Commit:
feat(dart/transform): Use the Dart transformer for benchmarks

Remove explicit generation of reflection information in benchmark code
and generate it with the transformer.
